File: 动态规划/背包问题/494.py
from typing import List


class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        Sum = 0
        for num in nums:
            Sum += num
        if (Sum-target)<0 or (Sum-target)%2!=0:
            return 0
        total = (Sum-target)//2
        dp = [[0 for _ in range(total+1)] for _ in range(len(nums)+1)]
        dp[0][0] = 1
        for i in range(1,len(nums)+1):
            num = nums[i-1]
            for j in range(total+1):
                if(j>=num):
                    dp[i][j] = dp[i-1][j]+dp[i-1][j-num]
                else:
                    dp[i][j] = dp[i-1][j]
        return dp[len(nums)][total]
        
nums = [1,1,1,1,1]
target = 3
obj = Solution()
print(obj.findTargetSumWays(nums,target))

File: 动态规划/背包问题/494.cpp
#include<vector>
using namespace std;
class Solution {
public:
    int findTargetSumWays(vector<int> &nums, int target)
    {
        int Sum = 0;
        for(auto num:nums)
            Sum += num;
        if(((Sum-target)<0)||(Sum-target)%2!=0)
            return 0;
        int total = (Sum-target)/2;
        vector<vector<int>>dp(nums.size()+1,vector<int>(total+1));
        dp[0][0] = 1;
        for(int i=1;i<nums.size()+1;i++){
            int num = nums[i-1];
            for(int j=0;j<total+1;j++){
                dp[i][j] = dp[i-1][j];
                if(j>=num)
                    dp[i][j]+=dp[i-1][j-num];
            }
        }
        return dp[nums.size()][total];
    }
};